### PR TITLE
Enable fusion mode to identify fusion transcript from fusion genome alignments

### PR DESCRIPTION
**Pull request by jonathangoeke**
Changes:
- the fusion mode is now available, which assigns read classes that align to multiple genes to a new combined fusion gene
- multi gene assignment is implemented in scoreReadClasses, which now uses the minimum distance threshold. This change will impact results in general beyond the fusion mode
- a new option is included to filter out subset transcripts if their relative read count compared to all superset transcripts is below a threshold
- the prepareAnnotations function now creates unique read and gene names from a GTF file
- the annotation objects now include a integerlist with equivalent classes for more efficient eqClass representation and faster computation. This is used for the subset calculation

### CODE REVIEW

**Review by andredsim — COMMENTED** [review]
Below I have the stats when I compared the following row data columns between the original and this PR. I included the results from when I first did the review and now, so the changes to assignGeneIds is in PR2. As you can see the first changes to the PR did not impact the gene assignments much, not the gene read proportions. However the changes in PR2 make much larger changes. 
One contributor is that there are 17854 read classes with NA gene id in PR2. I believe these are from non stranded, non overlapping read classes that I mention in a comment in the review. 


are equal? | Original vs PR1 | Original vs PR2 | PR1 vs PR2
-- | -- | -- | --
Read Count | 333021 | 333021 | 333021
Gene Read Prop | 329617 | 186363 | 187052
GENEID | 306140 | 294434 | 294386
novelGene | 332887 | 322251 | 322275

File: R/bambu-processReads_utilityConstructReadClasses.R [hunk 1]
```diff
@@ -356,23 +356,57 @@ initiateHitsDF <- function(hitsWithin, grangesReference, stranded) {
...
+      }
+  }
+  newGeneSet <- is.na(mcols(grl)$GENEID) & strandedRanges
+  newGeneIds <- assignGeneIdsNoReference(grl[newGeneSet])
+  mcols(grl)$GENEID[newGeneSet] <- newGeneIds
+  if(any(!strandedRanges)) {
+    mcols(grl)$GENEID[!strandedRanges] <- assignGeneIdsByReference(grl[!strandedRanges], 
+                                                                   grl[!is.na(mcols(grl)$GENEID)],
+                                                       min.exonOverlap = min.exonOverlap,
+                                                       fusionMode = FALSE)
```
[the remark this answers is not in the document]
Comment (jonathangoeke) [reply R/bambu-processReads_utilityConstructReadClasses.R:392]: fixed in [5e612a1](https://github.com/GoekeLab/bambu/pull/257/commits/5e612a1d30e95fd511a3d1da9f40e6bccf4fd755)

[the remark this answers is not in the document]
Comment (jonathangoeke) [reply R/bambu-processReads_utilityConstructReadClasses.R:395]: they should be assigned a novel gene id (unstranded), solved in  5e612a1d30e95fd511a3d1da9f40e6bccf4fd755

File: R/bambu-processReads_utilityConstructReadClasses.R [hunk 2]
```diff
@@ -356,23 +356,57 @@ initiateHitsDF <- function(hitsWithin, grangesReference, stranded) {
...
+    mcols(grl)$GENEID[strandedRanges] <- assignGeneIdsByReference(grl[strandedRanges], annotations,
+                                          min.exonOverlap = min.exonOverlap,
+                                          fusionMode = fusionMode) 
+    #iteratively assign gene ids for stranded granges
+    newGeneSet <- is.na(mcols(grl)$GENEID) & strandedRanges
+    referenceGeneSet <- !is.na(mcols(grl)$GENEID) & strandedRanges
+    mcols(grl)$GENEID[newGeneSet] <- assignGeneIdsByReference(grl[newGeneSet], grl[referenceGeneSet],
+                                                                  min.exonOverlap = min.exonOverlap,
+                                                                  fusionMode = FALSE)  # fusion assignment is only done based on original annotations
+    while(any(!is.na(mcols(grl)$GENEID[newGeneSet]))) {
```
[the remark this answers is not in the document]
Comment (jonathangoeke) [reply R/bambu-processReads_utilityConstructReadClasses.R]: - implemented a parameter for maximum iterations in d07c374eb722d5f7b8bca7d32b5e2dd5cae43c35
- a threshold using read count is not implemented, but the extendAnnotations uses complex filtering procedures that will handle these cases. Could be implemented as an additional feature though if that is an issue

**Pull request closed without merging**